### PR TITLE
Allow Dirackernel to use stateful property

### PR DESCRIPTION
**Pull request by jiangwen84**
I put the codes here and would like to hear any suggestions.   

Right now it has several limitations:
1. it should not work with adaptivity
2. it does not work with recover
3. When the number of quadrature points is changed (such as add new points or delete old points), all material property within this element will be initialized. For example, if one element has one point already and you want to add one more point to this element, the material property of these two points will be initialized. 

Otherwise, it should be working fine.

Tagging @dschwen @permcody @friedmud for suggestions.

Ref #7291

### CODE REVIEW

**Comment by dschwen** [comment]
Just to recap some of our discussion from yesterday: This will probably not make sense for 99% of DiracKernel applications (and those will not need this capability anyways). In a DK the number and locations and of active points is redetermined at every timestep. There is nothing that maps points from the current timestep to the point in the previous timestep. Except in XFEM (which is what this is for).

**Comment by jiangwen84** [comment]
I agree. This might be very specific to our own xfem application. :-)

**Comment by jiangwen84** [comment]
Seems it does work with recover. I might have to disable it in the test file. :-(

**Comment by friedmud** [comment]
Actually - it needs to work with recovery first. Please don't disable it in
the test.

**Comment by friedmud** [comment]
I'll try to take a look through this this morning...

File: test/tests/dirackernels/stateful_point_source/tests [hunk 1]
```diff
@@ -1,7 +1,8 @@
 [Tests]
   [./stateful_point_source_test]
-    type = 'RunException'
+    type = 'Exodiff'
     input = 'stateful_point_source.i'
-    expect_err = "Error: Stateful material properties not allowed"
+    exodiff = "out.e"
+    recover = false
```
Comment (friedmud) [inline test/tests/dirackernels/stateful_point_source/tests]: `recover = false` is not an option for something that is core functionality like this.  If the design can't be made to recover... then it needs a different design.

File: framework/src/materials/MaterialPropertyStorage.C [hunk 2]
```diff
@@ -350,6 +350,32 @@ MaterialPropertyStorage::swapBack(MaterialData & material_data, const Elem & ele
...
 
 bool
+MaterialPropertyStorage::needInitialized(unsigned int n_qpoints, const Elem & elem, unsigned int side)
+{
+  bool need = false;
+
+  // Has not been intialized
+  if (props()[&elem][side].size() == 0 || propsOld()[&elem][side].size() == 0) need = true;
+  if (hasOlderProperties())
+    if (propsOlder()[&elem][side].size() == 0) need = true;
```
Comment (friedmud) [inline framework/src/materials/MaterialPropertyStorage.C]: Please put the body of the `if` on its own line

File: framework/src/materials/MaterialPropertyStorage.C [hunk 3]
```diff
@@ -350,6 +350,32 @@ MaterialPropertyStorage::swapBack(MaterialData & material_data, const Elem & ele
...
+
+  // The number of q_points has changed
+  if (!need)
```
Comment (friedmud) [inline framework/src/materials/MaterialPropertyStorage.C]: The logic here is kind of bizarre and hard to follow.

Can you explain what you're trying to do here and maybe we can work on restructuring this.

File: framework/src/materials/MaterialPropertyStorage.C [hunk 4]
```diff
@@ -350,6 +350,32 @@ MaterialPropertyStorage::swapBack(MaterialData & material_data, const Elem & ele
...
+  {
+    for (unsigned int i=0; i < _stateful_prop_id_to_prop_id.size(); ++i)
+    {
+      if (props()[&elem][side][i]->size() != n_qpoints || propsOld()[&elem][side][i]->size() != n_qpoints) need = true;
```
Comment (friedmud) [inline framework/src/materials/MaterialPropertyStorage.C]: Body of `if` on own line

File: framework/src/materials/MaterialPropertyStorage.C [hunk 5]
```diff
@@ -350,6 +350,32 @@ MaterialPropertyStorage::swapBack(MaterialData & material_data, const Elem & ele
...
+      if (hasOlderProperties())
+        if (propsOlder()[&elem][side][i]->size() != n_qpoints) need = true;
```
Comment (friedmud) [inline framework/src/materials/MaterialPropertyStorage.C]: Body of `if` on own line

File: framework/src/materials/MaterialPropertyStorage.C [hunk 6]
```diff
@@ -253,14 +253,14 @@ MaterialPropertyStorage::initStatefulProps(MaterialData & material_data, const s
...
 
   // init properties (allocate memory. etc)
-  for (unsigned int i=0; i < _stateful_prop_id_to_prop_id.size(); ++i)
+  for (unsigned int i = 0; i < _stateful_prop_id_to_prop_id.size(); ++i)
   {
     // duplicate the stateful property in property storage (all three states - we will reuse the allocated memory there)
     // also allocating the right amount of memory, so we do not have to resize, etc.
-    if (props()[&elem][side][i] == NULL) props()[&elem][side][i] = material_data.props()[ _stateful_prop_id_to_prop_id[i] ]->init(n_qpoints);
-    if (propsOld()[&elem][side][i] == NULL) propsOld()[&elem][side][i] = material_data.propsOld()[ _stateful_prop_id_to_prop_id[i] ]->init(n_qpoints);
+    if (props()[&elem][side][i] == NULL || props()[&elem][side][i]->size() != n_qpoints) props()[&elem][side][i] = material_data.props()[ _stateful_prop_id_to_prop_id[i] ]->init(n_qpoints);
```
Comment (friedmud) [inline framework/src/materials/MaterialPropertyStorage.C]: Body of `if` on own line

File: framework/src/materials/MaterialPropertyStorage.C [hunk 7]
```diff
@@ -253,14 +253,14 @@ MaterialPropertyStorage::initStatefulProps(MaterialData & material_data, const s
...
+    if (propsOld()[&elem][side][i] == NULL || propsOld()[&elem][side][i]->size() != n_qpoints) propsOld()[&elem][side][i] = material_data.propsOld()[ _stateful_prop_id_to_prop_id[i] ]->init(n_qpoints);
```
Comment (friedmud) [inline framework/src/materials/MaterialPropertyStorage.C]: Same

File: framework/src/materials/MaterialPropertyStorage.C [hunk 8]
```diff
@@ -253,14 +253,14 @@ MaterialPropertyStorage::initStatefulProps(MaterialData & material_data, const s
...
     if (hasOlderProperties())
-      if (propsOlder()[&elem][side][i] == NULL) propsOlder()[&elem][side][i] = material_data.propsOlder()[ _stateful_prop_id_to_prop_id[i] ]->init(n_qpoints);
+      if (propsOlder()[&elem][side][i] == NULL || propsOlder()[&elem][side][i]->size() != n_qpoints) propsOlder()[&elem][side][i] = material_data.propsOlder()[ _stateful_prop_id_to_prop_id[i] ]->init(n_qpoints);
```
Comment (friedmud) [inline framework/src/materials/MaterialPropertyStorage.C]: Same

File: framework/src/base/FEProblem.C [hunk 9]
```diff
@@ -1820,20 +1828,28 @@ FEProblem::addMaterial(const std::string & mat_name, const std::string & name, I
       object_name = name + "_neighbor";
       MooseSharedPointer<Material> neighbor_material = _factory.create<Material>(mat_name, object_name, current_parameters, tid);
 
+      // dirac material
+      current_parameters.set<Moose::MaterialDataType>("_material_data_type") = Moose::DIRAC_MATERIAL_DATA;
+      current_parameters.set<bool>("_dirac") = true;
+      object_name = name + "_dirac";
+      MooseSharedPointer<Material> dirac_material = _factory.create<Material>(mat_name, object_name, current_parameters, tid);
```
Comment (friedmud) [inline framework/src/base/FEProblem.C]: This is a no-go.  It's adding a burden to every MOOSE solve out there... for the tiny possibility of doing something that has never even been allowed for the last 8 years.

If you want to enable DiracKernels to have some access to stateful material properties... you're going to need to do it "on demand" somehow...

File: framework/src/materials/MaterialPropertyStorage.C (hunk 3 above)
Comment (jiangwen84) [inline framework/src/materials/MaterialPropertyStorage.C]: I understand that the material stateful properties are all initialized in ```FEProblem``` during initial setup. However, for ```DiracKernel```, it can not be done during initial setup (at that time, the system does not know any ```DiracKernel``` points) but has to do every time step if necessary. The code here will return true if the material property on this element is never initialized/allocated in ```MaterialPropertyStorage``` or if the number of points is changed on this element.

File: framework/src/base/FEProblem.C (hunk 9 above)
Comment (friedmud) [inline framework/src/base/FEProblem.C]: Actually - I could be convinced that this is ok.

@idaholab/moose-developers what do you guys think?  Is this reasonable?

A `Material` might hold quite a lot of data in it... even if storage for its properties is never allocated...

File: framework/src/base/FEProblem.C [hunk 10]
```diff
@@ -1882,6 +1898,31 @@ FEProblem::reinitMaterials(SubdomainID blk_id, THREAD_ID tid, bool swap_stateful
...
 
 void
+FEProblem::reinitMaterialsDirac(SubdomainID blk_id, THREAD_ID tid, bool swap_stateful)
+{
+  if (_all_materials[Moose::DIRAC_MATERIAL_DATA].hasActiveBlockObjects(blk_id, tid))
+  {
+    const Elem * & elem = _assembly[tid]->elem();
+    unsigned int n_points = _assembly[tid]->qRule()->n_points();
+
+    if (_dirac_material_data[tid]->nQPoints() != n_points)
```
Comment (friedmud) [inline framework/src/base/FEProblem.C]: What happens if the number of points in the elem stays the same... but they MOVE within the element?  The stateful data is no longer valid... but this PR will still happily return old values...

Comment (jiangwen84) [inline framework/src/base/FEProblem.C]: That is correct. Currently I do not have a good solution for that.

I understand that the ```MaterialPropertyStorage``` does not store any information about the quadrature coordinates but just the quadrature index. So the book keeping will be complicated.

Moving a point within a element is one situation. Another is to add and delete the same number of points. For that case, the return value is not correct neither. :-(

File: test/tests/dirackernels/stateful_point_source/tests (hunk 1 above)
Comment (jiangwen84) [inline test/tests/dirackernels/stateful_point_source/tests]: I was looking into the recover issue and I think it fails because the material property storage is not initialized before the restart starts to run. This comes back to the basic design question, such as when should we init/reinit/resize material property value for these ```Dirac``` points. If there are any good solution, I would like to give it a shot. :-)

**Comment by WilkAndy** [comment]
The following is probably irrelevant here, but i wanted to mention it just in case.  About a year ago a DiracKernel didn't cache stuff (via addPoint(x,y,z,i) and currentPointCachedID) and my simulations ran at least 2x slower because each linear iteration MOOSE had to search around in the mesh for points, etc.  As i said, perhaps this is irrelevant, but if you have tonnes of Dirac Points, these geometrical searches can completely dominate compute time.

**Comment by permcody** [comment]
Closing this PR due to inactivity.
